In [1]:
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

#### Create a model that will use mpg as the response variable, and the other 10 columns of data as potential predictor variables. It is assumed that all 10 predictors don't really belong in the model.



In [2]:
Fitall <- lm(mpg~.,data=mtcars)

#### mpg is your y, or response variable, and . means "all." Instead of saying mpg ~ ., you could have listed out all 10 predictor variables, as in "mpg ~ cyl + disp + hp + drat + ...". It is much easier to use the . rather than list them all out.

In [8]:
formula(Fitall)

mpg ~ cyl + disp + hp + drat + wt + qsec + vs + am + gear + carb

#### using formula shows, obvi the formula. important later on.

In [9]:
summary(Fitall)



Call:
lm(formula = mpg ~ ., data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.4506 -1.6044 -0.1196  1.2193  4.6271 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) 12.30337   18.71788   0.657   0.5181  
cyl         -0.11144    1.04502  -0.107   0.9161  
disp         0.01334    0.01786   0.747   0.4635  
hp          -0.02148    0.02177  -0.987   0.3350  
drat         0.78711    1.63537   0.481   0.6353  
wt          -3.71530    1.89441  -1.961   0.0633 .
qsec         0.82104    0.73084   1.123   0.2739  
vs           0.31776    2.10451   0.151   0.8814  
am           2.52023    2.05665   1.225   0.2340  
gear         0.65541    1.49326   0.439   0.6652  
carb        -0.19942    0.82875  -0.241   0.8122  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.65 on 21 degrees of freedom
Multiple R-squared:  0.869,	Adjusted R-squared:  0.8066 
F-statistic: 13.93 on 10 and 21 DF,  p-value: 3.793e-07

#### Although the overall p value at the bottom is significant, none of the individual predictors are, meaning that the model is probably not a good fit; the number of variables has just inflated the p value to the point of being significant, and what the independent variables are does not really matter.

Once you have created a new model after backward elimination, you can compare the above model summary with all 10 predictors against what R has helped determine is the best fit.



In [10]:
step(Fitall, direction="backward")

Start:  AIC=70.9
mpg ~ cyl + disp + hp + drat + wt + qsec + vs + am + gear + carb

       Df Sum of Sq    RSS    AIC
- cyl   1    0.0799 147.57 68.915
- vs    1    0.1601 147.66 68.932
- carb  1    0.4067 147.90 68.986
- gear  1    1.3531 148.85 69.190
- drat  1    1.6270 149.12 69.249
- disp  1    3.9167 151.41 69.736
- hp    1    6.8399 154.33 70.348
- qsec  1    8.8641 156.36 70.765
<none>              147.49 70.898
- am    1   10.5467 158.04 71.108
- wt    1   27.0144 174.51 74.280

Step:  AIC=68.92
mpg ~ disp + hp + drat + wt + qsec + vs + am + gear + carb

       Df Sum of Sq    RSS    AIC
- vs    1    0.2685 147.84 66.973
- carb  1    0.5201 148.09 67.028
- gear  1    1.8211 149.40 67.308
- drat  1    1.9826 149.56 67.342
- disp  1    3.9009 151.47 67.750
- hp    1    7.3632 154.94 68.473
<none>              147.57 68.915
- qsec  1   10.0933 157.67 69.032
- am    1   11.8359 159.41 69.384
- wt    1   27.0280 174.60 72.297

Step:  AIC=66.97
mpg ~ disp + hp + drat + wt + qsec + am


Call:
lm(formula = mpg ~ wt + qsec + am, data = mtcars)

Coefficients:
(Intercept)           wt         qsec           am  
      9.618       -3.917        1.226        2.936  


#### the lower the AIC, the better the fit is for the predictors for the model. Scroll to the bottom, and none in carats is at the top of the list, and no more predictors can be removed to lower the AIC.

#### AIC stands for Akaike Information Criteria. It was invented by a Japanese statistician named Hirotugu Akaike in the 1970's. The AIC is a measure of relative quality of a model. However, it doesn't tell you if a model is any good or not. It simply helps you to compare two or more models against each other. When comparing two models, the model with the smaller AIC is generally accepted to be the better of the two models. In the case of the mtcars data, the start AIC of 70.9 is for the original model with all ten predictors.

#### The table has been sorted from smallest AIC at the top to largest AIC at the bottom. The takeaway from this table is as follows: Of all the models with just 9 predictors, it would appear that the model that excludes cyl is the best, because it has the smallest AIC.

Before moving on, there are a couple important points to make:

You don't really know if the model with cyl removed is any good or not, you just know that it is better than all the other models with a single predictor removed.

You might assume at this point that all of the terms above the row with "" could easily be removed. Since individually removing each of those rows makes for a better model, why not just remove all of them in one fell swoop, just to save time? This is actually an erroneous thought process. The power of stepwise regression is that each iteration evaluates a model in the presence or absence of the other terms in the model. This will be illustrated shortly based on the fact that at this first iteration, it might appear that the only terms that belong in the model are wt and am.

#### above refers to collinearity and the relationship or influence one variable may have on another variable.

#### Have you been paying attention to the models below the <none> row in each table? Remember, the interpretation of those rows is that if those variables get removed, the model will actually be a bit worse than the current model. Have you noticed that at each iteration, the variables below the <none> line are not always the same variables?

### WHY

#### If you look for a description of the mtcars dataset on the web, you will discover that wt is the weight of the car in thousands of pounds, qsec is the time it takes for the car to cover 1/4 mile from rest, and am is a coded variable for the type of transmission (0 for automatic, and 1 for manual).

#### According to the output table, the model that includes all three of those variables is better than any of the three models that omits one of those three variables. You are done with creating a model using backward elimination. If you haven't reached this conclusion on your own, don't worry, R has reached it for you.This output tells you that the best model for the mtcars data uses wt, qsec, and am to predict the mpg for a car, using the following equation:

mpg = 9.618 + (-3.917) _ wt + 1.226 _ qsec + 2.936 \* am

Pay attention to the signs of the coefficients for a moment:

The coefficient of wt is (-3.917). What that means is as long as everything else is equal, an increase of a car's weight by 1000 lbs will lead to an estimated decrease in the car's mpg by about 3.9 miles per gallon of gas.
The coefficient of qsec is 1.226. Everything else being equal, as a car's qsec increases, so does the mpg. Does that make sense? Well, yes - if the qsec is larger, that means the car is slower off the start. If you know anything about cars, more power is usually associated with more speed, which is usually associated with worse gas mileage. In this case, you are being told that less power leads to higher gas mileage - for every additional second it takes to run a quarter mile, the gas mileage is improved by about 1.2 miles per gallon. It is pretty much an accepted fact that gutless cars get better gas mileage than fast cars.
The coefficient for the transmission type indicates that all other things being equal, two identical cars where one is an automatic transmission and one is a manual transmission will have different miles per gallon ratings. The manual transmission should get around 3 mpg better mileage than the automatic transmission. 

#### So now you have a model. Is it any good? You know it is the best model you could find using backward elimination, but is it any good?

You can determine this by creating a new model with just those variables and then looking at the summary. You will create a new function called fitsome, which will just have the terms you are interested in:

In [11]:
fitsome = lm(mpg~am + qsec + wt, data=mtcars)

In [12]:
summary(fitsome)


Call:
lm(formula = mpg ~ am + qsec + wt, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.4811 -1.5555 -0.7257  1.4110  4.6610 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   9.6178     6.9596   1.382 0.177915    
am            2.9358     1.4109   2.081 0.046716 *  
qsec          1.2259     0.2887   4.247 0.000216 ***
wt           -3.9165     0.7112  -5.507 6.95e-06 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 2.459 on 28 degrees of freedom
Multiple R-squared:  0.8497,	Adjusted R-squared:  0.8336 
F-statistic: 52.75 on 3 and 28 DF,  p-value: 1.21e-11


#### There are a few things to note here from the above output:

There are coefficients for just the three predictor variables, as well as an intercept.
There is a multiple R2 = 0.8497. The practical interpretation of this is that the model explains 84.97% of the variation in the mpg variable, and there is another 15.03% of the variation that can be chalked up to noise or random error.
There is an adjusted R2 = 0.8336. The adjustment is a modification that is supposed to take into account the number of terms in the model. For your purpose, you are more interested in the adjusted R2 than the multiple R2.
There is an F-statistic (52.75 with 3 and 28 degrees of freedom) and a p value of 0.0000000000121 (this is represented in scientific notation in R) which indicates that the model is better than no model at all, because the p value is less than 0.05.
So, have you really gained anything by doing the backward elimination? To the untrained eye, one might say "no gain." The R2 values are very similar, and both models are statistically significant. However, to the trained eye, you have adhered to the "Principle of Parsimony" and come up with an equally good (maybe marginally better) model that is much simpler. Further, now all of your individual predictors are significant! Mission accomplished.

